juliac: log entrypoints and nonstandard typedefs

timholy opened the pull request:
To facilitate creating bindings for other programming languages, this adds logging to the `juliac` buildscript, optionally outputting a logfile that describes the entrypoints (signatures and return types) and nonstandard typedefs that appear in the entrypoint specifications.

My Makefile skills are rudimentary at best, so please do check that aspect with particular care.

timholy committed:
juliac: log entrypoints and nonstandard typedefs

To facilitate creating bindings for other programming languages, this
adds logging to the `juliac` buildscript, optionally outputting a
logfile that describes the entrypoints (signatures and return types) and
nonstandard typedefs that appear in the entrypoint specifications.

timholy committed:
minor Makefile improvements

timholy committed:
Nicer printing

- remove mangling in favor of native Julia printing
- include argnames in methods
- support and test name argument to ccallable

JeffBezanson commented on contrib/juliac/juliac-buildscript.jl:
```diff
@@ -27,6 +27,30 @@ if Base.JLOptions().trim != 0
     include(joinpath(@__DIR__, "juliac-trim-base.jl"))
 end
 
+const C_friendly_types = Union{    # a few of these are redundant to make it easier to maintain
```
This should be a `Set{Any}`.

timholy committed:
Use IdSet instead of Union

timholy committed:
Try using the library

timholy commented:
OK, this is getting more ambitious but running into issues. I'm now trying to actually use the library and not seeing the names I need:
```
tim@kenzo:~/src/julia/test/trimming$ make JULIA='../../julia --startup-file=no' BIN=.
../../julia --startup-file=no -t 1 -J /home/tim/src/julia/usr/bin/../lib/julia/sys.so --startup-file=no --history-file=no --output-o hello-o.a --output-incremental=no --strip-ir --strip-metadata --experimental --trim /home/tim/src/julia/usr/bin/../share/julia/juliac/juliac-buildscript.jl /home/tim/src/julia/test/trimming/hello.jl --output-exe true
gcc -march=native -mtune=native -m64 -o hello -Wl,--whole-archive hello-o.a -Wl,--no-whole-archive   -std=gnu11 -I'/home/tim/src/julia/usr/include/julia' -fPIC  -lm -L'/home/tim/src/julia/usr/lib' -Wl,--export-dynamic -L'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib' -ljulia -ljulia-internal
../../julia --startup-file=no -t 1 -J /home/tim/src/julia/usr/bin/../lib/julia/sys.so --startup-file=no --history-file=no --project=/home/tim/src/julia/test/trimming -e "using Pkg; Pkg.instantiate()"
../../julia --startup-file=no -t 1 -J /home/tim/src/julia/usr/bin/../lib/julia/sys.so --startup-file=no --history-file=no --project=/home/tim/src/julia/test/trimming --output-o basic_jll-o.a --output-incremental=no --strip-ir --strip-metadata --experimental --trim /home/tim/src/julia/usr/bin/../share/julia/juliac/juliac-buildscript.jl /home/tim/src/julia/test/trimming/basic_jll.jl --output-exe true
gcc -march=native -mtune=native -m64 -o basic_jll -Wl,--whole-archive basic_jll-o.a -Wl,--no-whole-archive   -std=gnu11 -I'/home/tim/src/julia/usr/include/julia' -fPIC  -lm -L'/home/tim/src/julia/usr/lib' -Wl,--export-dynamic -L'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib' -ljulia -ljulia-internal
../../julia --startup-file=no -t 1 -J /home/tim/src/julia/usr/bin/../lib/julia/sys.so --startup-file=no --history-file=no --output-o libsimple-o.a --output-incremental=no --strip-ir --strip-metadata --experimental --trim /home/tim/src/julia/usr/bin/../share/julia/juliac/juliac-buildscript.jl /home/tim/src/julia/test/trimming/libsimple.jl --output-lib true ./bindinginfo_libsimple.log
gcc -march=native -mtune=native -m64 -shared -o ./libsimple.so ./libsimple-o.a -lm -L'/home/tim/src/julia/usr/lib' -Wl,--export-dynamic -L'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib' -ljulia -ljulia-internal
gcc -march=native -mtune=native -m64 -I. -I/home/tim/src/julia/test/trimming -I/home/tim/src/julia/usr/bin/../lib -o capplication capplication.c -lm -L'/home/tim/src/julia/usr/lib' -Wl,--export-dynamic -L'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib/julia' -Wl,-rpath,'/home/tim/src/julia/usr/lib' -ljulia -ljulia-internal  -Wl,-rpath=. -L. -lsimple   -std=gnu11 -I'/home/tim/src/julia/usr/include/julia' -fPIC
/usr/bin/ld: /tmp/cczir9hy.o: in function `main':
capplication.c:(.text+0x8f): undefined reference to `copyto_and_sum'
/usr/bin/ld: capplication.c:(.text+0xf4): undefined reference to `countsame'
collect2: error: ld returned 1 exit status
make: *** [Makefile:58: capplication] Error 1
```

JeffBezanson commented on contrib/juliac/juliac-buildscript.jl:
```diff
@@ -27,6 +27,30 @@ if Base.JLOptions().trim != 0
+        T.name.module === Core && error("invalid type for juliac: ", T) # exclude internals (they may change)
+        push!(types, T)
+    end
+    for list in (T.parameters, fieldtypes(T))
```
I think this will only apply to `Ptr`. `Array` was added to `jl_type_mappable_to_c` but I'm not sure that's a good idea --- it expects a julia object reference so it's not really usable by a typical C program.

timholy committed:
Remove Array from C-friendly types

timholy committed:
Get the shared library working

timholy committed:
Clean up after trimming tests

JeffBezanson commented on contrib/juliac/juliac-buildscript.jl:
```diff
@@ -73,6 +96,43 @@ let mod = Base.include(Main, ARGS[1])
+    if length(ARGS) >= 4
+        logfile = ARGS[4]
+        iotmp = IOBuffer()
+        open(logfile, "w") do io
```
Would be good to move this to a function.

timholy committed:
Update to JuliaLibWrapping .h file

timholy committed:
Split `write_logfile` into separate function

topolarity commented on contrib/juliac/juliac-buildscript.jl:
```diff
@@ -27,6 +27,62 @@ if Base.JLOptions().trim != 0
+    return types
+end
+
+function write_logfile(io::IO)
```
My KISS instinct would be to use a standard serialization format like JSON to print out the type information

Also do we need the `offset` info? No problem with adding it so I'm mostly asking for understanding - IIUC the offsets should be implied by the C ABI on a particular platform

timholy replied:
Totally happy to receive help! The more hands the better.

topolarity replied:
Some changes at https://github.com/JuliaInterop/JuliaLibWrapping.jl/pull/2

it will need some more tweaks + new tests next week, but all tests are passing for now

topolarity commented on contrib/juliac/juliac-buildscript.jl:
```diff
@@ -73,6 +129,14 @@ let mod = Base.include(Main, ARGS[1])
     end
+
+    # Export info about entrypoints and structs needed to create header files
+    if length(ARGS) >= 4
```
Can we make this a `--export-abi build.json` (or other name - open to bikeshedding) instead of a positional arg?

topolarity commented on contrib/juliac/juliac-buildscript.jl:
```diff
@@ -27,6 +27,62 @@ if Base.JLOptions().trim != 0
+])
+
+function recursively_add_types!(types::Base.IdSet{DataType}, @nospecialize(T::DataType))
+    if T ∉ C_friendly_types
+        if T <: Ptr
+            return recursively_add_types!(types, T.parameters[1])
+        end
+        T.name.module === Core && error("invalid type for juliac: ", T) # exclude internals (they may change)
+        push!(types, T)
+    end
```
```suggestion
    while T <: Ptr
        T = T.parameters[1] # unwrap Ptr{...}
    end
    if T.name.module === Core && T ∉ C_friendly_types
        error("invalid type for juliac: ", T) # exclude internals (they may change)
    end
    push!(types, T)
```

Mostly just to make the logic more direct, but there is a slight behavioral difference - if we add an `isbits struct` to `C_friendly_types`, this will make sure it still gets added to `types` for translation.

topolarity reviewed (commented):
Thanks for taking the initiative on this, this is lovely!

Some thoughts on the implementation, but I am definitely in favor of the feature 👍

topolarity commented on test/trimming/libsimple.jl:
```diff
@@ -0,0 +1,44 @@
+    return sum(to)
+end
+
+Base.@ccallable function countsame(list::Ptr{MyTwoVec}, n::Int32)::Int32
```
Two slightly advanced (but important) features:
  - `Ref{...}` should be supported in addition to `Ptr{...}`
  - function pointers are just `Ptr{Cvoid}` ABI-wise but for C to be able to call them, we do need to export type definitions for them

gbaraldi replied:
Ref is a weird case, plain C code can't make a Ref, but embedding of julia can make one

vtjnash replied:
FWIW, ccallable never supported any of those "advanced" scenarios that cfunction supports

timholy committed:
Update contrib/juliac/juliac-buildscript.jl

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>

topolarity committed:
Export ABI metadata in JSON format

With luck, this makes this information easier to consume downstream (and
to expand on in the future).

Also fixes a couple of obscure bugs that would cause mis-behavior in the
presence of recursive types, or which could unintentionally mangle types
in the event of sub-string collisions.

topolarity committed:
Move ABI export functionality to `abi_export.jl`

topolarity committed:
Print field / argument info more compactly

topolarity committed:
Add `--export-abi` arg to `juliac.jl`

timholy reviewed (commented):
This looks great, many thanks!

topolarity commented on contrib/juliac/abi_export.jl:
```diff
@@ -0,0 +1,217 @@
+    Int8, Int16, Int32, Int64, UInt8, UInt16, UInt32, UInt64, Float32, Float64, Bool,
+    Cvoid, Cint, Cshort, Clong, Cuint, Cushort, Culong, Cssize_t, Csize_t,
+    Cchar, Cwchar_t, Cstring, Cwstring,
+    RawFD,
```
I think we can leave it

Downstream consumers (e.g. https://github.com/JuliaInterop/JuliaLibWrapping.jl/) should error if they aren't equipped to handle a custom `primitive` type, but the format also now includes a lot of extra attributes for these so that you can pick a native C type that matches the ABI (i.e. the number of bits / alignment / signed-ness, etc.)

topolarity committed:
Add extra metadata for `primitive` types

topolarity committed:
Minor fix-up for makefile paths

topolarity committed:
Merge remote-tracking branch 'julialang/master' into teh/juliac_headerlogs

topolarity committed:
Update header file to latest from `JuliaLibWrapping.jl`

topolarity committed:
Adjust to `Core.MethodTable` -> `Core.methodtable`

topolarity committed:
Minor fix-up for makefile paths

topolarity committed:
Update `test/trimming/trimming.jl` for JSON ABI format

timholy commented on contrib/juliac/abi_export.jl:
```diff
@@ -0,0 +1,219 @@
+    type_ids::IdDict{Any,Int}
+end
+
+function escape_string_json(s::AbstractString)
```
IIUC there isn't a "meaningful" test of this escaping.

timholy commented on contrib/juliac/abi_export.jl:
```diff
@@ -0,0 +1,219 @@
+            print(ctx.io, " " ^ (indent + 4))
+            print(ctx.io, "{")
+            print(ctx.io, " \"name\": ", escape_string_json(argnames[i]), ",")
+            print(ctx.io, " \"type\": ", ctx.type_ids[sig.parameters[i]])
```
Should this be `typeid` or `type_id` instead of just `type`? I don't feel strongly about this, but I suspect it might be clearer that it's a reference within the json file.

timholy commented on contrib/juliac/abi_export.jl:
```diff
@@ -0,0 +1,219 @@
+            println(ctx.io, i == length(sig.parameters) ? " }" : " },")
+        end
+        indented_println("  ],")
+        indented_println("  \"returns\": { \"type\": ", ctx.type_ids[rt], " }")
```
likewise here

timholy commented on contrib/juliac/abi_export.jl:
```diff
@@ -0,0 +1,219 @@
+    let indented_println(args...) = println(ctx.io, " " ^ indent, args...)
+        indented_println("{")
+        indented_println("  \"symbol\": ", escape_string_json(symbol), ",")
+        indented_println("  \"name\": ", escape_string_json(name), ",")
```
Maybe call this `julia_signature` instead of `name`? The latter seems too generic.

topolarity replied:
I intentionally kept this a bit generic to (mildly) discourage parsing of the signature

We can allow that of course, but it means we have to maintain an external signature parser and reckon with some outstanding issues, like that `isbits` objects can appear in types and have no canonical printing (https://github.com/JuliaLang/julia/issues/58756)

timholy reviewed (commented):
This looks very good. Only small quibbles, then I think this is ready to merge.

topolarity committed:
Rename `type` to `type_id`

topolarity committed:
Add recursive type tests

vtjnash commented on contrib/juliac/abi_export.jl:
```diff
@@ -0,0 +1,219 @@
+    print(ctx.io, " \"name\": ", field_name_json(dt, field), ",")
+    print(ctx.io, " \"type_id\": ", type_id, ",")
+    print(ctx.io, " \"offset\": ", desc.offset)
+    print(ctx.io, " }")
```
Can you add `Base.isfieldatomic` as an optional field

timholy committed:
Update contrib/juliac/abi_export.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

timholy committed:
Add isptr and isfieldatomic

timholy committed:
Merge branch 'master' into teh/juliac_headerlogs

timholy committed:
Fix spacing

timholy committed:
Debug tests

timholy commented:
Not quite sure how to fix the test failure here...I can't even seem to run these tests locally anymore.

timholy committed:
Remove project

timholy committed:
Remove module qualification

timholy committed:
fix tests

timholy commented:
Locally this now passes. Will merge soon unless there are other concerns. (I think the issue about combining JSON escaping across stdlibs etc is better done as a separate PR.)

timholy committed:
Add all reqd packages to Project

timholy committed:
Instantiate prior to trimming

timholy committed:
add a comment
